uid/src/main/java/com/tomtom/kotlin/uid/Uid.kt

```diff
@@ -0,0 +1,225 @@
+                UUID_MAX_LENGTH + "], but is " + uuid.length + ", uuid=" + uuid
+        }
+
+        // Check dashes.
+        for (i in UUID_DASH_POS) {
+            if (length < i + 1 || uuid[i] != UUID_DASH) {
+
+                // This will throw an IllegalArgumentException if it went wrong.
+                this.uuid = UUID.fromString(uuid).toString().toLowerCase()
+                return
+            }
+        }
+
+        // Make sure the UUID is lowercase.
+        this.uuid = uuid.toLowerCase()
+
+        // UUID seems to be OK.
+        kotlin.require(onlyContainsValidUUIDCharacters(this.uuid)) {
+            "Incorrect UUID format, uuid=$uuid"
+        }
+    }
+
+    /**
+     * Instantiates an id with a [UUID].
+     *
+     * @param uuid Existing [UUID].
+     */
+    private constructor(uuid: UUID) {
+        this.uuid = uuid.toString()
+    }
+
+    /**
+     * Returns hex string representation of this Uid. Opposite of [fromHexString].
+     * This representation is shorthand, it does not include dashes.
+     *
+     * @return Hex string representation of id, exactly 32 characters long.
+     */
+    fun toHexString(): String {
+        val uuid2 = UUID.fromString(uuid)
+        val msb = java.lang.Long.toHexString(uuid2.mostSignificantBits)
+        val lsb = java.lang.Long.toHexString(uuid2.leastSignificantBits)
+        val sb = StringBuilder()
+        for (i in 1..16 - msb.length) {
+            sb.append('0')
+        }
+        sb.append(msb)
+        for (i in 1..16 - lsb.length) {
+            sb.append('0')
+        }
+        sb.append(lsb)
+        val value = sb.toString()
+        assert(value.length == 32)
+        return value
+    }
+
+    /**
+     * Method converts given String representation to Uid and compares it with this instance.
+     * A String value of "0-0-0-0-0" would match a UID of "00000-0000-0000-000000000-00" or so.
+     *
+     * @param uid String representation od Uid.
+     * @return True in case String representation matches instance. False otherwise.
+     */
+    fun matchesFromString(uid: String): Boolean {
+        return this == fromString<T>(uid)
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is Uid<*>) return false
+        if (uuid != other.uuid) return false
+        return true
+    }
+
+    override fun hashCode(): Int {
+        return uuid.hashCode()
+    }
+
+    /**
+     * Returns string representation of this Uid. Opposite of [fromString].
+     *
+     * @return String representation of id.
+     */
+    override fun toString(): String {
+        return uuid
+    }
+
+    companion object {
+        private const val serialVersionUID = 1L
+        private const val UUID_DASH = '-'
+        private const val UUID_MIN_LENGTH = 9
+        private const val UUID_MAX_LENGTH = 36
+        private val UUID_DASH_POS = intArrayOf(8, 13, 18, 23)
+
+        /**
+         * Return whether an ID is a valid UUID.
```

Comment:
This may be used in input validation, for example, or database field validation. The constructor throws, which is inconvenient in such cases, so this wraps it. The method `fromStringIfValid` isn't that clear, either.

Reply:
Additional problem: what type should be returned. The signature should then be `fromStringifValid<T>(...)` which makes it even less clear. I'll leave it for now. We'll add it as a feature if we need it.

Reply:
The `@return` documentation and that of `fromString` does recommend the duplicate execution though. Should we not add a `<T> fromStringIfValid` to be able to recommend more efficient usage? (After all, efficiency is what this tool is about.)

Reply:
OK